packages/flutter_tools/lib/src/migrate/migrate_utils.dart

```diff
@@ -0,0 +1,362 @@
+    return result.exitCode == 0;
+  }
+
+  /// Runs `flutter pub upgrade --major-revisions`.
+  Future<void> flutterPubUpgrade(String workingDirectory) async {
+    final List<String> cmdArgs = <String>['flutter', 'pub', 'upgrade', '--major-versions'];
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory, allowReentrantFlutter: true);
+    checkForErrors(result, commandDescription: cmdArgs.join(' '));
+  }
+
+  /// Runs `./gradlew tasks` in the android directory of a flutter project.
+  Future<void> gradlewTasks(String workingDirectory) async {
+    final String baseCommand = _platform.isWindows ? 'gradlew.bat' : './gradlew';
+    final List<String> cmdArgs = <String>[baseCommand, 'tasks'];
+    final RunResult result = await _processUtils.run(cmdArgs, workingDirectory: workingDirectory);
+    checkForErrors(result, commandDescription: cmdArgs.join(' '));
+  }
+
+  /// Verifies that the RunResult does not contain an error.
+  ///
+  /// If an error is detected, the error can be optionally logged or exit the tool.
+  ///
+  /// Passing -1 in allowedExitCodes means all exit codes are valid.
+  bool checkForErrors(
+    RunResult result, {
+    List<int> allowedExitCodes = const <int>[0],
+    String? commandDescription,
+    bool exit = true,
+    bool silent = false
+  }) {
+    if ((result.exitCode != 0 && !allowedExitCodes.contains(result.exitCode)) && !allowedExitCodes.contains(-1)) {
+      if (!silent) {
+        _logger.printError('Command encountered an error with exit code ${result.exitCode}.');
+        if (commandDescription != null) {
+          _logger.printError('Command:');
+          _logger.printError(commandDescription, indent: 2);
+        }
+        _logger.printError('Stdout:');
+        _logger.printError(result.stdout, indent: 2);
+        _logger.printError('Stderr:');
+        _logger.printError(result.stderr, indent: 2);
+      }
+      if (exit) {
+        throwToolExit('Command failed with exit code ${result.exitCode}', exitCode: result.exitCode);
+      }
+      return false;
+    }
+    return true;
+  }
+
+  /// Returns true if the file does not contain any git conflit markers.
+  bool conflictsResolved(String contents) {
+    if (contents.contains('>>>>>>>') && contents.contains('=======') && contents.contains('<<<<<<<')) {
+      return false;
+    }
+    return true;
+  }
+}
+
+/// Defines the classification of difference between files.
+enum DiffType {
+  modification,
+  addition,
+  deletion,
+  ignored,
+}
+
+/// Tracks the output of a git diff command or any special cases such as addition of a new
+/// file or deletion of an existing file.
+class DiffResult {
+  DiffResult({
+    required this.diffType,
+    this.diff,
+    this.exitCode,
+  });
+
+  /// The diff string output by git.
+  final String? diff;
+
+  final DiffType diffType;
+
+  /// The exitcode of the command. This is zero when no diffs are found.
```

Comment:
What does null exit code mean?

Reply:
I redefined the behavior around this member. It is null when no command is run to obtain the DiffResult. When DiffType is `command`, this is populated.